{
 "metadata": {
  "name": ""
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%matplotlib inline\n",
      "from rnapredUtils import *\n",
      "from rnapredCV import *\n",
      "\n",
      "expr_file = '/home/sofiakp/projects/roadmap/rna/Dec13/medusaInputs/noDupGenes/57epigenomes_mergedPc_expr.txt'\n",
      "gene_name_file = '/home/sofiakp/projects/roadmap/rna/Dec13/medusaInputs/noDupGenes/57epigenomes_mergedPc_genes.txt'\n",
      "expt_name_file = '/home/sofiakp/projects/roadmap/rna/Dec13/medusaInputs/noDupGenes/57epigenomes_mergedPc_exptnames.txt'\n",
      "out_dir = '/home/sofiakp/scg3_lab/users/sofiakp/roadmap/rnapred/feat_mat/'\n",
      "local_out_dir = '/home/sofiakp/projects/roadmap/rnapred/feat_mat/'\n",
      "gene_file = '/home/sofiakp/projects/Anshul/data/tss/gencode19/gencode.v19.annotation.noM.genes.bed'\n",
      "all_expt_name_file = '/home/sofiakp/scg3_lab/users/sofiakp/roadmap/cell_line_ids.txt'\n",
      "gc_cont_file = '/home/sofiakp/scg3_lab/users/sofiakp/data/gencode/v19/gencode.v19.annotation.noM.genes.firstTSS.1k.gc'\n",
      "array_file = '/home/sofiakp/projects/roadmap/rna/exon_arrays/gene_avg_expr.npz'\n",
      "\n",
      "if not os.path.isdir(out_dir):\n",
      "    os.mkdir(out_dir)\n",
      "\n",
      "feat_names = ['H3K4me1.firstTSS.1k', 'H3K4me3.firstTSS.1k', \n",
      "              'H3K27me3.firstTSS.1k', 'H3K36me3.firstTSS.1k', 'H3K9me3.firstTSS.1k',\n",
      "              'H3K27me3.body.1k', 'H3K36me3.body.1k', 'H3K9me3.body.1k',\n",
      "              'H3K4me1.lastTES.1k', 'H3K4me3.lastTES.1k', \n",
      "              'H3K27me3.lastTES.1k', 'H3K36me3.lastTES.1k', 'H3K9me3.lastTES.1k',\n",
      "              'H3K27me3.firstTSS.1k', 'H3K36me3.firstTSS.1k', 'H3K9me3.firstTSS.1k',\n",
      "              'H3K27me3.body.1k', 'H3K36me3.body.1k', 'H3K9me3.body.1k',\n",
      "              'H3K27me3.lastTES.1k', 'H3K36me3.lastTES.1k', 'H3K9me3.lastTES.1k']\n",
      "\n",
      "out_feat_names = feat_names[0:13]\n",
      "out_feat_names.extend(['broad.' + f for f in feat_names[13:]])\n",
      "\n",
      "signal_dirs = ['/home/sofiakp/scg3_lab/users/sofiakp/roadmap/rnapred/signalAtPeaks/narrowPeaks/rowPeaks/pval/'] * 13\n",
      "signal_dirs.extend(['/home/sofiakp/scg3_lab/users/sofiakp/roadmap/rnapred/signalAtPeaks/broadPeaks/pval/'] * 9)\n",
      "assert(len(feat_names) == len(signal_dirs))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Read data \n",
      "gene_names, short_expt_names, all_expt_names = read_rna_pred_data(\\\n",
      "                    expr_file, gene_name_file, expt_name_file, gc_cont_file, array_file,\n",
      "                    gene_file, all_expt_name_file, feat_names, signal_dirs,\n",
      "                    out_feat_names, out_dir, overwrite = False)\n",
      "\n",
      "np.savetxt(os.path.join(out_dir, 'all_expt_names.txt'), all_expt_names, fmt = ['%s'])\n",
      "train_expt_names = []\n",
      "for ex in short_expt_names:\n",
      "    if os.path.isfile(os.path.join(out_dir, ex + '_feat_mat.npz')):\n",
      "        data = np.load(os.path.join(out_dir, ex + '_feat_mat.npz'))\n",
      "        if 'y' in data:\n",
      "            train_expt_names.append(ex)\n",
      "np.savetxt(os.path.join(out_dir, 'train_expt_names.txt'), train_expt_names, fmt = ['%s'])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Compute correlation between exon arrays and RNA-seq\n",
      "\n",
      "out_cells = []\n",
      "cors = []\n",
      "outfile = os.path.join(out_dir, 'array_rnaseq_cor.txt')\n",
      "\n",
      "with open(os.path.join(out_dir, 'train_expt_names.txt'), 'r') as infile:\n",
      "    for line in infile:\n",
      "        ex = line.strip()\n",
      "        feat_mat_file = os.path.join(out_dir, ex + '_feat_mat.npz')\n",
      "        if not os.path.exists(feat_mat_file):\n",
      "            continue\n",
      "        data = np.load(feat_mat_file)\n",
      "        ex_ind = np.argwhere(np.array(data['feat_names']) == 'array_expr')[0][0]\n",
      "        out_cells.append(ex)\n",
      "        cors.append(np.corrcoef(data['y'], data['feat'][:, ex_ind])[0][1])\n",
      "        data.close()\n",
      "\n",
      "with open(outfile, 'w') as f:\n",
      "    for n, c in zip(out_cells, cors):\n",
      "        f.write('\\t'.join([n, str(c)]) + '\\n')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Plot the results of cross validation for each experiment\n",
      "# (and for the merged data if available).\n",
      "\n",
      "from rnapredPlots import *\n",
      "\n",
      "method = 'rf'\n",
      "cv_dir = os.path.join(out_dir, 'cv', method)\n",
      "\n",
      "# Plots will go here.\n",
      "if not os.path.isdir(os.path.join(cv_dir, 'plots')):\n",
      "    os.mkdir(os.path.join(cv_dir, 'plots'))\n",
      "\n",
      "# File with maximum (average) R-squared for each experiment.\n",
      "summary_file = os.path.join(cv_dir, 'cv_summary.txt')\n",
      "\n",
      "with open(os.path.join(out_dir, 'train_expt_names.txt'), 'r') as infile:\n",
      "    expt_names = [line.strip() for line in infile.readlines()]\n",
      "expt_names.append('merged')\n",
      "\n",
      "with open(summary_file, 'w') as outfile:\n",
      "    for ex in expt_names:\n",
      "        cv_file = os.path.join(cv_dir, ex + '_cv.npz')\n",
      "        if not os.path.exists(cv_file):\n",
      "            continue\n",
      "        cv_res = read_cv_res(cv_file)\n",
      "        \n",
      "        if method == 'rf':\n",
      "            df = forest_cv_to_pandas_df(cv_res)\n",
      "        else:\n",
      "            df = logLass_cv_to_pandas_df(cv_res)\n",
      "        for ext in ['pdf', 'png']:\n",
      "            plot_cv_r2(df, os.path.join(cv_dir, 'plots', ex + '_cv.' + ext))\n",
      "        \n",
      "        df2 = df[['title', 'r2']]\n",
      "        df2 = pd.groupby(df2, 'title').mean()\n",
      "        max_r2 = np.max(df2['r2'])\n",
      "        max_params = df2[df2['r2'] == max_r2]['r2'].index\n",
      "        outfile.write('\\t'.join([ex, str(max_r2), ','.join(max_params)]) + '\\n')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 5
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Plot the results of cross validation for merged experiments\n",
      "# Now each experiment is a test set with the rest of the experiments \n",
      "# being the training set. Will create one plot with the R-squared \n",
      "# for each set of parameters.\n",
      "\n",
      "import rnapredPlots\n",
      "from rnapredPlots import *\n",
      "\n",
      "method = 'logLasso'\n",
      "cv_dir = os.path.join(out_dir, 'cv', method)\n",
      "\n",
      "if not os.path.isdir(os.path.join(cv_dir, 'plots')):\n",
      "    os.mkdir(os.path.join(cv_dir, 'plots'))\n",
      "\n",
      "# File with average R-squared for each set of parameters.\n",
      "summary_file = os.path.join(cv_dir, 'merged_cv_summary.txt')\n",
      "\n",
      "with open(os.path.join(out_dir, 'train_expt_names.txt'), 'r') as infile:\n",
      "    expt_names = [line.strip() for line in infile.readlines()]\n",
      "\n",
      "if method == 'rf':\n",
      "    df_fun = forest_cv_to_pandas_df\n",
      "else:\n",
      "    df_fun = logLass_cv_to_pandas_df\n",
      "\n",
      "# Concatenate the results from all the experiments.\n",
      "for expt_idx, ex in enumerate(expt_names):\n",
      "    cv_res = read_cv_res(os.path.join(cv_dir, 'merged_no_' + ex + '_cv.npz'))\n",
      "    if expt_idx == 0:\n",
      "        df = df_fun(cv_res, cv = 1)[['title', 'r2']]\n",
      "    else:\n",
      "        df = pd.concat([df, df_fun(cv_res, cv = 1)[['title', 'r2']]])\n",
      "    df.index = np.arange(df.shape[0])\n",
      "\n",
      "for ext in ['pdf', 'png']:\n",
      "    plot_cv_r2(df, os.path.join(cv_dir, 'plots', 'merged_hoo_cv.' + ext))\n",
      "df2 = pd.groupby(df, 'title').mean()\n",
      "df2.to_csv(os.path.join(cv_dir, 'merged_hoo_cv_summary.txt'), sep = '\\t')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 7
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Plot the results of learning on other experiments \n",
      "# (i.e. what happens when the training data come from an entirely \n",
      "# different cell type)\n",
      "\n",
      "from rnapredPlots import *\n",
      "\n",
      "method = 'rf'\n",
      "# The optimal parameters might differ, depending on whether we train on each experiment\n",
      "# separately or on all of them merged.\n",
      "# param_str = ['C1_a0.0001', 'C1_a100', 'C1_a0.0001']\n",
      "param_str = ['t200_s5_t200_s5', 't200_s100_t200_s5', 't200_s5_t200_s5']\n",
      "if method == 'logLasso':\n",
      "    njobs = 1\n",
      "else:\n",
      "    njobs = 6\n",
      "model_dirs = [os.path.join(out_dir, 'cv', method, p) for p in param_str]\n",
      "\n",
      "with open(os.path.join(out_dir, 'train_expt_names.txt'), 'r') as infile:\n",
      "    expt_names = [line.strip() for line in infile.readlines()]\n",
      "r2_avg = r2_trained_on_avg_rest(expt_names, local_out_dir, model_dirs[0], njobs = njobs)\n",
      "r2_rest = r2_trained_on_rest(expt_names, local_out_dir, model_dirs[1], njobs = njobs)\n",
      "r2_corr = r2_trained_on_most_corr(expt_names, local_out_dir, model_dirs[2], njobs = njobs)\n",
      "\n",
      "titles = ['Average', 'Merged rest', 'Best']\n",
      "titles = [t + ' (' + p + ')' for (t, p) in zip(titles, param_str)]\n",
      "\n",
      "outfile = os.path.join(out_dir, 'cv', method, 'plots', 'merged_comp.txt')\n",
      "with open(outfile, 'w') as f:\n",
      "    f.write('{}\\t{:.4f}\\n{}\\t{:.4f}\\n{}\\t{:.4f}\\n'.format(titles[0], np.mean(r2_avg), \n",
      "                                                          titles[1], np.mean(r2_rest), \n",
      "                                                          titles[2], np.mean(r2_corr)))\n",
      "\n",
      "r2_avg.extend(r2_rest)\n",
      "r2_avg.extend(r2_corr)\n",
      "\n",
      "ncv = len(r2_rest)\n",
      "types = [titles[0]] * ncv\n",
      "types.extend([titles[1]] * ncv)\n",
      "types.extend([titles[2]] * ncv)\n",
      "df = pd.DataFrame({'r2':r2_avg, 'title':types})\n",
      "\n",
      "for ext in ['pdf', 'png']:\n",
      "    outfile = os.path.join(out_dir, 'cv', method, 'plots', 'merged_comp.' + ext)\n",
      "    plot_cv_r2(df, outfile, height = 120, max_width = 80, fsize = 12, xlab = '')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 6
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Plot feature importance or coefficients for different features across models.\n",
      "\n",
      "from rnapredPlots import *\n",
      "\n",
      "method = 'rf'\n",
      "param_str = 't200_s5_t200_s5'\n",
      "model_dir = os.path.join(local_out_dir, 'cv', method, param_str)\n",
      "# Model files are expected to be <model_dir>/<pref><expt>_model.pkl\n",
      "pref = ''\n",
      "\n",
      "plot_dir = os.path.join(out_dir, 'cv', method, 'plots')\n",
      "if not os.path.isdir(plot_dir):\n",
      "    os.mkdir(plot_dir)\n",
      "\n",
      "with open(os.path.join(out_dir, 'train_expt_names.txt'), 'r') as infile:\n",
      "    expt_names = [line.strip() for line in infile.readlines()]\n",
      "\n",
      "for ext in ['pdf', 'png']:\n",
      "    plot_coef(out_dir, model_dir, expt_names, pref, height = 130, fsize = 12,\n",
      "              outfile = os.path.join(plot_dir, param_str + '_' + pref + 'coef.' + ext))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 3
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Test the effect of number of trees in R-squared\n",
      "\n",
      "import rnapredPlots\n",
      "from rnapredPlots import *\n",
      "\n",
      "param_str = 't200_s100_t200_s5'\n",
      "model_dir = os.path.join(local_out_dir, 'cv', 'rf', param_str)\n",
      "\n",
      "ntrees = [10, 20, 50, 100, 150, 200]\n",
      "\n",
      "with open(os.path.join(out_dir, 'train_expt_names.txt'), 'r') as infile:\n",
      "    expt_names = [line.strip() for line in infile.readlines()]\n",
      "    \n",
      "r2 = []\n",
      "# Concatenate the results from all the experiments.\n",
      "for expt_idx, ex in enumerate(expt_names):\n",
      "    feat_mat_file = os.path.join(out_dir, ex + '_feat_mat.npz')\n",
      "    model_file = os.path.join(model_dir, 'merged_no_' + ex + '_model.pkl')\n",
      "    if not os.path.exists(feat_mat_file) or not os.path.exists(model_file):\n",
      "        continue\n",
      "    model = read_model(model_file)\n",
      "    feat, y, feat_names, gene_names = read_feat_mat(feat_mat_file)\n",
      "    ntrees_tmp, r2_tmp = r2_per_tree_size(model, feat, y, ntrees)\n",
      "    assert(len(ntrees) == len(ntrees_tmp))\n",
      "    r2.extend(r2_tmp)\n",
      "\n",
      "nexpt = len(r2) / len(ntrees)\n",
      "df = pd.DataFrame({'r2':r2, 'trees':ntrees * nexpt})\n",
      "\n",
      "r_df = com.convert_to_r_dataframe(df)\n",
      "fsize = 12\n",
      "gp = ggplot2.ggplot(r_df) + ggplot2.aes_string(x = 'factor(trees)', y = 'r2') + \\\n",
      "    ggplot2.geom_boxplot() + ggplot2.scale_y_continuous('R-squared') + \\\n",
      "    ggplot2.scale_x_discrete('Number of trees') + ggplot2.theme_bw() + \\\n",
      "    ggplot2.theme(**{'axis.text.x':ggplot2.element_text(size = fsize),\n",
      "                     'axis.text.y':ggplot2.element_text(size = fsize)})\n",
      "for ext in ['pdf', 'png']:\n",
      "    outfile = os.path.join(out_dir, 'cv', 'rf', 'plots', param_str + '_merged_hoo_ntree_comp.' + ext)\n",
      "    ro.r.ggsave(filename = outfile, plot = gp, width = 100, height = 100, unit = 'mm')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "/usr/local/lib/python2.7/dist-packages/numpy/lib/utils.py:1132: DeprecationWarning: The compiler package is deprecated and removed in Python 3.x.\n",
        "  import compiler\n"
       ]
      }
     ],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "model = read_model('/home/sofiakp/projects//roadmap/rnapred/feat_mat/cv/rf/t200_s100_t200_s5/merged_no_E128_model.pkl')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 30
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def kl(p, q):\n",
      "    m = (p+q)/2\n",
      "    return np.sum(np.where(p == 0, 0, p * np.log2(p/m) + q * np.log2(q/m)))\n",
      "\n",
      "p = np.array([0.0, 0.0, 0.0, 1])\n",
      "q = np.array([0.1, 0.0, 0.1, 0.8])\n",
      "print kl(p, q)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "(expt_names, expt_name_idx) = get_expt_map(cel_map_file, cel_names)\n",
      "summary = plot_cels(expr, expt_names, expt_name_idx, cel_names, '/home/sofiakp/projects/roadmap/rna/exon_arrays/corr_plots')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "Caching the list of root modules, please wait!\n",
        "(This will only be done once - type '%rehashx' to reset cache!)\n",
        "\n"
       ]
      }
     ],
     "prompt_number": 51
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print r_df"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "             r2 trees\n",
        "0  0.707033....    10\n",
        "1  0.716127....    20\n",
        "2  0.724028....    50\n",
        "3  0.724887....   100\n",
        "4  0.725459....   150\n",
        "5  0.725912....   200\n",
        "6  0.734511....    10\n",
        "7  0.743789....    20\n",
        "8  0.749259....    50\n",
        "9  0.750465....   100\n",
        "10 0.751767....   150\n",
        "11 0.752208....   200\n",
        "\n"
       ]
      }
     ],
     "prompt_number": 44
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}